### PR TITLE
[TFLu] Update initialization of 'input_dims' in EvalQuantizedInt8() method of Fully Connected kernel (CMSIS-NN)

### LINKED ISSUE #41816: TFLM 2.3 fails assert in CMSIS-NN fully_connected.cc
@tensorflow/micro

**System information**
- Host OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Verified on Mbed OS 5.13
- TensorFlow installed from (source or binary): Source
- Tensorflow version (commit SHA if source): 0c43ad89f81b22d81d1894f5b53f9fbdda0b738a
- Target platform (e.g. Arm Mbed OS, Arduino Nano 33 etc.): Mbed OS (ST IoT Discovery Kit)

**Describe the problem**

We have a very simple fully connected network (33x20x10x4 neurons) that fails an assert when being invoked with CMSIS-NN kernels here (in **fully_connected.cc**):

```
    cmsis_nn_dims input_dims;
    input_dims.n = batches;
    input_dims.h = input_shape.Dims(1);
    input_dims.w = input_shape.Dims(2); // <-- here
```

Here `input_shape` is the input layer (33 neurons) which does not have this dimension (`size_` is 2). Setting `input_dims.w` and `input_dims.c` to `1` solves the issue.

**Please provide the exact sequence of commands/steps when you ran into the problem**

See above. Attached the tflite model.
[ei-continuous-gestures-nn-classifier-tensorflow-lite-int8-quantized-model.lite.zip](https://github.com/tensorflow/tensorflow/files/4987903/ei-continuous-gestures-nn-classifier-tensorflow-lite-int8-quantized-model.lite.zip)

### PR DESCRIPTION
**Pull request by biagiom**
This is a new pull request based on the comments of #41285, which aims to fix initialization of `cmsis_nn_dims input_dims`.

Fixes #41816

### CODE REVIEW

**Review by felix-johnny — COMMENTED** [review]
Looks good!

**Comment by felix-johnny** [comment]
@njeffrie Could you have a look at this PR, please? I recently came across an issue for a network where a reshape reduced the dimensions to 2 and caused a crash. Would be good to have this fix in.

**Review by advaitjain — COMMENTED** [review]
Looks good to me. One request is to have a unittest that fails without this PR, but passes with it.

**Comment by gbaned** [comment]
@biagiom  Can you please check @advaitjain's comments and keep us posted ? Thanks!

**Comment by biagiom** [comment]
Hi @advaitjain,
sorry for my late response but I was very busy in the last two weeks.
As for your comment, could you please provide more details (and if possible help me) about how to build the unittests ?

Moreover, I just tested the `hello_world` example (compiled on the latest source code of TFLite) on a [DISCO-L475VG-IOT01A](https://os.mbed.com/platforms/ST-Discovery-L475E-IOT01A/) and it returns the same exception discussed in my previous PR (#41285). However, with the fix introduced in this PR, the example runs without any issue.
I would just point out that the model related to the `hello_world` example uses only 3 fully-connected layers without applying flattening and when the model is executed, the input tensors to the fully-connected layers seem to have only 2 dimensions.
I also tried MobileNet v1 (tf.keras.applications.MobileNet) and I got the same exception when executing the fully-connected layer.
Maybe they can be considered as "a kind of test cases" that might help you to further analyze this issue.
In any case, I think my PR works both when the input is a bi-dimensional tensor and a 4-dim tensor.
Finally, are you able to reproduce the issue on the `hello_world` example ? Please make sure to add the flag `__ARM_FEATURE_DSP=1` when compiling the example in order to use the CMSIS-NN optimized kernels.

Best regards,
Biagio.

**Comment by advaitjain** [comment]
I'd like to see a new test case in micro/kernels/fully_connected_test.cc that exercises this code.

Apologies for not looking at this closely before -- it would additionally be good to also explain why this new mapping between input_shape and the cmsis-nn input_dims is indeed generic and not model specific.

Tagging @freddan80 for their thoughts.

**Comment by freddan80** [comment]
Hello @biagiom and thanks for your contribution! As @advaitjain mentions, adding a testcase covering the input that caused the issue would be great (in micro/kernels/fully_connected_test.cc). From https://github.com/tensorflow/tensorflow/issues/41816, I guess the testcase would be similar to an existing one, but have size_=2 according to: _Here input_shape is the input layer (33 neurons) which does not have this dimension (size_ is 2)._. That way we avoid this issue coming back. It's enough to add a test there and run it on host for now I believe. You can try run it using Renode on an emulated Cortex-M, but I don't think it's necessary. We run all these unit tests regularly on Cortex-M models. 

Regarding `__ARM_FEATURE_DSP=1`, it's a predefined macro set by the compiler, so you should not set it. More info here: https://developer.arm.com/documentation/dui0774/g/Other-Compiler-specific-Features/Predefined-macros?lang=en. Basically, as soon as you compile using a Cortex-M processor target, this flag is set automatically. Let me know if that doesn't work properly for you, but perhaps we can take that discussion in a separate Github issue or via email, not to derail this thread.

Both me and @felix-johnny are ooo for some days, so I'll loop in @jenselofsson

Hope this helps!
/Fredrik

**Comment by gbaned** [comment]
@biagiom Can you please check @freddan80's comments and keep us posted ? Thanks!

**Comment by biagiom** [comment]
Hi @advaitjain and @freddan80,
first of all sorry for my late response and above all thanks for your comments and tips for the issue.

I have done some further tests and analyzed in more detail my contribution and I would like to point out the following considerations:

1. First I will start from the @advaitjain comment:
> Apologies for not looking at this closely before -- it would additionally be good to also explain why this new mapping between input_shape and the cmsis-nn input_dims is indeed generic and not model specific.

With respect to the current implementation, the new mapping I introduced with my PR is more generic because it is able to manage both 2D and 4D input tensors. In fact, as for 2D input tensors, `input_shape.Dims(2);` and `input_shape.Dims(3);` would generate an exception. In particular, if you look at [lines 196 and 197 of /tensorflow/lite/kernels/internal/types.h](https://github.com/tensorflow/tensorflow/blob/cc30260fe86eb00693dc953655611f068b5c4546/tensorflow/lite/kernels/internal/types.h#L196), they will check if the dimension `i` is greater or equal than 0 and lower than `size_`, the number of dimensions of the input tensor. If any of the two conditions does not hold, both the checks will evaluate to `TFLITE_ASSERT_FALSE`.

2. As for the test case, I have created a new example called [fully_connected_cmsis_test in my TensorFlow fork (branch cmsis_fc_test)](https://github.com/biagiom/tensorflow/tree/cmsis_fc_test/tensorflow/lite/micro/examples/fully_connected_cmsis_test) that contains the same code you can find in micro/kernels/fully_connected_test.cc as it tests fully-connected op using both 2D and 4D input tensors and several data formats (float, int8, uint8).
I compiled the example with the following commands and tested on my [DISCO-L475VG-IOT01A](https://os.mbed.com/platforms/ST-Discovery-L475E-IOT01A/):
```
$ make -f tensorflow/lite/micro/tools/make/Makefile TARGET=mbed TAGS="cmsis-nn" generate_fully_connected_cmsis_test_mbed_project
$ cd tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/fully_connected_cmsis_test/mbed
$ mbed config root .
$ mbed deploy
$ mbed config GCC_ARM_PATH "~/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin"
$ mbed compile -m DISCO_L475VG_IOT01A -t GCC_ARM
$ cp ./BUILD/DISCO_L475VG_IOT01A/GCC_ARM/mbed.bin /media/biagio/DIS_L4IOT
```
As for results, all the test cases are passed without any issue with the new mapping (the one described in my PR).
Another important point is that it still works without any issue also if using the old mapping even for the test cases that make use of a 2D input tensor.
At that point I was a bit confused as I expected that the test cases using a 2D input tensor would fail.
So, I investigated in detail the source code and discovered that this strange behavior is given by the use of `NDEBUG` macro:
if you look at [line 30 of /tensorflow/lite/kernels/op_macros.h](https://github.com/tensorflow/tensorflow/blob/23cb51dfb64f863d4e5d470160d6b44f1ffdb992/tensorflow/lite/kernels/op_macros.h#L60), you can note that if `NDEBUG` is defined, then `TFLITE_ASSERT_FALSE` is defined as `(static_cast<void>(0))`. This means that any assertion based on `TFLITE_ASSERT_FALSE` (like `TFLITE_DCHECK_GE` and `TFLITE_DCHECK_LT` used by `Dims()` method) will not have any effect ! So, even though `EvalQuantizedInt8()` method uses `input_shape.Dims(2)` and `input_shape.Dims(3)` with 2D input tensors, there isn't any exception.
In my TensorFlow fork I have modified the [fully_connected CMSIS-NN implementation](https://github.com/biagiom/tensorflow/blob/cmsis_fc_test/tensorflow/lite/micro/kernels/cmsis-nn/fully_connected.cc) to make some checks at runtime and (above all) test this behavior of `NDEBUG`. I also modified the [op_macros.h](https://github.com/biagiom/tensorflow/blob/cmsis_fc_test/tensorflow/lite/kernels/op_macros.h) file in order to undefine the `NDEBUG` macro (see the notes in the comments).
Now, if you clone my repo and try to compile the `fully_connected_cmsis_test` or `hello_world` example (any example that use fully_connected op), the execution will stop in a loop because now, since `TFLITE_ASSERT_FALSE` is defined with `TFLITE_ABORT`, when `EvalQuantizedInt8()` calls `input_shape.Dims(2)`, `TFLITE_DCHECK_LT(i, size_);` evaluates to `TFLITE_ABORT`.

Anyway, it seems that it doesn't matter how the fields `h`, `w` and `c` of `input_dims` are initialized because if you look at `arm_fully_connected_s8` to which `input_dims` is passed to, the only field that seems to be used is the `n` field (batch size).
However I would prefer that this issue related to the use of `NDEBUG` will be handled in a different way as it may cause unexpected behaviors (see #41816).

I hope to be clear in my explanation.
Moreover, I think that we don't need any additional test unit in `micro/kernels/fully_connected_test.cc` as all the test cases can be used to test the code I want to merge (maybe we just need to update the [BUILD](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/kernels/BUILD) file to ensure we use the CMSIS-NN implementation instead of the default one).

What do you think about the previous points ? Is there anything of my observations that I missed or is wrong ? 

Best regards,
Biagio.

**Comment by freddan80** [comment]
Hi @biagiom

Thanks for your detailed analysis! And you are correct. The tests that already exist for fully connected should have caught this. It should have failed on `TFLITE_DCHECK_LT`, as you say: https://github.com/tensorflow/tensorflow/blob/cc30260fe86eb00693dc953655611f068b5c4546/tensorflow/lite/kernels/internal/types.h#L197
@advaitjain , do you know why it doesn't hit the assert? It seems like we can input any number, for example:
```
  const RuntimeShape input_shape = tflite::micro::GetTensorShape(input);
  static int test = input_shape.Dims(15);
```
and it won't hit the assert. For all targets and all tests. I have a quick look and `_size` seem ok. I'd prefer this to be handled in a separate ticket.

> Moreover, I think that we don't need any additional test unit in micro/kernels/fully_connected_test.cc as all the test cases can be used to test the code I want to merge (maybe we just need to update the BUILD file to ensure we use the CMSIS-NN implementation instead of the default one).

Agree, we can go ahead and merge this PR imo.

**Comment by felix-johnny** [comment]
@biagiom Thanks for the analysis. What we see here is the typical difference between release(-DNDEBUG, asserts are disabled) and debug(-DDEBUG, asserts are enabled) builds. @freddan80 that is why we don't hit the asserts.

 I can understand why a release build option is used as without it, the performance of TFLu ref kernels will fall sharply especially because of the use of asserts(its a comparison statement from the compilers point of view) in the inlined Offset() routine of the core loop as well as in the MultiplyByQuantizedMultiplier() routines of common functions. 

 Maybe building it with -DDEBUG for the unit tests only would be an option. But, that is a topic outside the purview of this PR, and I think this PR can be merged as it is.

**Comment by advaitjain** [comment]
Apologies for the delay here.

With d6138980ec98e0634002ca4b23bea6061e46fde4, the default makefile build will have these asserts.

One more clarification for the fix from this PR:

A different TFLM user encountered the same issue and proposed a different fix which involved changing the input_dims assignment to:
```cc
input_dims.n = batches;
input_dims.h = input_shape.Dims(1);
input_dims.w = input_shape.DimensionsCount() >= 3 ? input_shape.Dims(2) : 1;
input_dims.c = input_shape.DimensionsCount() >= 4 ? input_shape.Dims(3) : 1;
```

And this PR is suggesting:
```
input_dims.n = batches;
input_dims.h = 1;
input_dims.w = 1;
input_dims.c = accum_depth;
```

What is the preferred solution here? It appears to me that the first solution is semantically better but I'm curious what everyone else thinks.

**Comment by felix-johnny** [comment]
@advaitjain While the first solution gives a "I know what I am doing here" vibe, I look at the dimension checks as something that has already been done during the model/layer creation and used to update accum_depth(a.k.a filter filter dimension). 

 The second solution is a simpler one that builds up on known information and is clearer to read. For those reasons, I prefer solution 2.

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
Sounds good, let's merge this PR.

**Comment by biagiom** [comment]
Hi @advaitjain,
sorry for my late comment.

Even though I think that the first solution is more elegant since it checks the dimensions of the input tensor, the second one (the one I proposed) it simpler and easier to read so I agree with @felix-johnny that we can merge the commit as it is.

Anyway, thanks to all for your support @freddan80, @felix-johnny, @advaitjain !

Best regards,
Biagio.
